File: solutions/1-two-sum/two-sum.go
package main

import "fmt"

func twoSum(nums []int, target int) []int {

	numMap := make(map[int]int)

	for idx := 0; idx < len(nums); idx++ {

		value := nums[idx]
		diff := target - value

		if diffIdx, found := numMap[diff]; found {
			return []int{diffIdx, idx}
		}

		numMap[value] = idx
	}

	return []int{}
}

func twoSum1(nums []int, target int) []int {

	for outerIdx := 0; outerIdx < len(nums); outerIdx++ {

		for innerIdx := outerIdx + 1; innerIdx < len(nums); innerIdx++ {

			if nums[outerIdx]+nums[innerIdx] == target {
				return []int{outerIdx, innerIdx}
			}
		}
	}

	return []int{}
}

func main() {
	nums := []int{2, 7, 11, 15}
	target := 9
	fmt.Println(twoSum(nums, target))
}

File: solutions/1-two-sum/two-sum.ts
function twoSum(nums: number[], target: number): number[] {
  if (!nums || !nums.length || typeof target != "number") {
    return [];
  }

  const myMap = new Map<number, number>();

  for (let i = 0; i < nums.length; i++) {
    const currentValue = nums[i];
    const remainder = target - currentValue;

    if (myMap.has(remainder)) {
      const remainderIndex = myMap.get(remainder)!;
      return [remainderIndex, i];
    }

    myMap.set(currentValue, i);
  }

  return [];
}

function twoSum1(nums: number[], target: number): number[] {
  if (!nums || !nums.length || typeof target != "number") {
    return [];
  }

  for (let outerIndex = 0; outerIndex < nums.length; outerIndex++) {
    for (
      let innerIndex = outerIndex + 1;
      innerIndex < nums.length;
      innerIndex++
    ) {
      const value = nums[outerIndex] + nums[innerIndex];

      if (value === target) {
        return [outerIndex, innerIndex];
      }
    }
  }

  return [];
}
